### PR TITLE
Lambda improvements

### LINKED ISSUE #1338: lambda doesn't support get function by name and version
When calling the get_function API and specifying the version in the qualifier parameter, the call throws a ClientError('An error occurred (404) when calling the GetFunction operation: ') 

function_data = client.get_function(FunctionName="my_func", Qualifier="$LATEST")  

Looking at the source code, this occurs because the version gets concatenated to the function name.

### PR DESCRIPTION
**Pull request by terrycain**
Fixes #1338 

Completely revamped how lambdas are stored, now we have lambda publishing working and aliasing should be trivial to add, attempted to spam `weakref` everywhere but I have a feeling my debugger was messing with the refs so theres more boilerplate code that I would of expected.

Done the majority of this whilst sleep deprived, @JackDanger keep an eye out for the spelling mistakes ;-).

Also cleaned up the responses.py, made use of a json property and backend property, I think I inadvertently fixed some region malarkey when in server mode too.

### CODE REVIEW

File: moto/awslambda/models.py [hunk 1]
```diff
@@ -161,7 +167,7 @@ def replace_adapter_send(*args, **kwargs):
             'VpcConfig', {'SubnetIds': [], 'SecurityGroupIds': []})
 
         # auto-generated
```
Comment (JackDanger) [inline moto/awslambda/models.py]: Is this comment still true?

File: moto/awslambda/models.py [hunk 2]
```diff
@@ -203,11 +209,15 @@ def replace_adapter_send(*args, **kwargs):
                 self.code_size = key.size
                 self.code_sha_256 = hashlib.sha256(key.value).hexdigest()
 
-        self.function_arn = 'arn:aws:lambda:{}:123456789012:function:{}'.format(
-            self.region, self.function_name)
+        self.function_arn = make_function_arn(self.region, ACCOUNT_ID, self.function_name, version)
```
Comment (JackDanger) [inline moto/awslambda/models.py]: nice.

File: moto/awslambda/models.py [hunk 3]
```diff
@@ -515,23 +636,31 @@ def send_message(self, function_name, message):
...
+        self._functions[function_name][-1].invoke(json.dumps(event), {}, {})
         pass
 
     def list_tags(self, resource):
         return self.get_function_by_arn(resource).tags
 
     def tag_resource(self, resource, tags):
-        self.get_function_by_arn(resource).tags.update(tags)
+        fn = self.get_function_by_arn(resource)
+        if fn:
```
Comment (JackDanger) [inline moto/awslambda/models.py]: minor suggestion: I'd personally prefer a guard clause here that early-returns on `if not fn` so the rest of the method isn't as indented. Same as the following method.

File: moto/awslambda/models.py [hunk 4]
```diff
@@ -446,41 +455,153 @@ def create_from_cloudformation_json(cls, resource_name, cloudformation_json,
         return LambdaVersion(spec)
 
 
+class LambdaStorage(object):
```
Comment (JackDanger) [inline moto/awslambda/models.py]: This is great

File: tests/test_awslambda/test_lambda.py [hunk 5]
```diff
@@ -317,30 +317,25 @@ def test_get_function():
...
-            "Timeout": 3,
-            "Version": '$LATEST',
-            "VpcConfig": {
-                "SecurityGroupIds": [],
-                "SubnetIds": [],
-            }
-        },
-        'ResponseMetadata': {'HTTPStatusCode': 200},
-    })
+    result['Code']['Location'].should.equal('s3://awslambda-{0}-tasks.s3-{0}.amazonaws.com/test.zip'.format(_lambda_region))
```
Comment (JackDanger) [inline tests/test_awslambda/test_lambda.py]: Nice. Way less brittle

File: moto/awslambda/models.py (hunk 1 above)
Comment (terrycain) [reply moto/awslambda/models.py]: Hmm, possibly, will investigate.

File: moto/awslambda/models.py (hunk 3 above)
Comment (terrycain) [reply moto/awslambda/models.py]: Ahh, now we get down to personal preference :wink:. But I agree, less indentation is nice, and its an easy fix

**Pull request merged**